Add react native tutorial (#32)
jellyfish-dev/jellyfish-docs: docs/tutorials/react-native.mdx

gracefully leave the room, close the server connection and go back to the
Connect screen.

For leaving the room and closing the server connection you can use the `cleanUp` method from the `useJellyfishClient()` hook.

After leaving the room we go back to the Connect screen.

```tsx title="/screens/Room.tsx"
// ...
import {
  usePeers,
  VideoRendererView,
  // highlight-next-line
  useJellyfishClient,
} from "@jellyfish-dev/react-native-client-sdk";
// highlight-next-line
import { InCallButton } from "@jellyfish-dev/react-native-jellyfish-components";

function RoomScreen({ navigation }): JSX.Element {
  const peers = usePeers();
  // highlight-start
  const { cleanUp } = useJellyfishClient();

  const onDisconnectPress = () => {
    cleanUp();
    navigation.goBack();
  };
  // highlight-end

  return (
    <View style={styles.container}>
      <VideosGrid
        tracks={peers.map((peer) => peer.tracks[0]?.id).filter((t) => t)}
      />
      // highlight-start
      <InCallButton
        type="disconnect"
        iconName="phone-hangup"
        onPress={onDisconnectPress}
      />
      // highlight-end
    </View>
  );
}

// ...
```

> ## Summary

Reviewer: I had to install:

```
npm install @react-navigation/native-stack
```

> ## Summary

Reviewer: Please add section about:

```
npm start
npm run ios
```

> ## Summary

Reviewer: And still don't work. This is my outcome:
https://github.com/kamil-stasiak/JellyfishReactNativeMinimalExample

Could you help me with that?

<img width="647" alt="image" src="https://github.com/jellyfish-dev/jellyfish-docs/assets/19818474/7afe3532-e27b-4dda-947f-a92cf0279e25">
<img width="386" alt="image" src="https://github.com/jellyfish-dev/jellyfish-docs/assets/19818474/9ed2727f-efc5-452a-a5c2-c49439771bd1">

<img width="388" alt="image" src="https://github.com/jellyfish-dev/jellyfish-docs/assets/19818474/0454d61a-0f35-418a-84ec-73cd6bf755df">

---

import { ImageWrapper } from "./ImageWrapper";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import JellyfishArchitecture from "./common/_jellyfish-architecture.mdx";
import StartingJellyfishBackend from "./common/_starting-jellyfish-backend.mdx";
import StartingJellyfishDashboard from "./common/_starting-dashboard.mdx";

# React Native Minimal Working Example

## What you'll learn

This tutorial will guide you through creating your first React Native / Expo
project which uses Jellyfish client. By the end of the tutorial, you'll have a
working application that connects to the front-end dashboard using WebRTC
technology and streams and receives camera tracks.

<ImageWrapper url={'/img/tutorials/rn_full_app.gif'}/>

You can check out the finished project [here](https://github.com/jellyfish-dev/jellyfish-clients-tutorials/tree/main/JellyfishDashboardRN).

## What do you need

- a little bit of experience in creating apps with React Native and/or Expo -
  refer to the [React Native
  Guide](https://reactnative.dev/docs/getting-started) or [Expo
  Guide](https://docs.expo.dev/) to learn more

## Jellyfish architecture

<JellyfishArchitecture />

> ### Start the Jellyfish backend

Reviewer 2: `Starting JF backend` and `Starting JF dashboard` doesn't fit into  `Jellyfish architecture`, does it? I would suggest:

1. What you'll learn?
2. What do you need?
3. Jellyfish archutecture (?)
4. Setup 
  a) Start JF
  b) Start JF Dashboard
  c) Create React Native/Expo project
  d) Add dependencies
  e) Permissions configuration
  d) Add components library (maybe should be in `Add dependencies`)?
5. Screens
6. ConnectScreen
7. RoomScreen
 
  
  
Also, avoid using the term `Jellyfish backend` in the context of `Jellyfish Media Server.` `Jellyfish backend` is the `Jellyfish Dashboard` in this tutorial.

### Start the dashboard web front-end

<StartingJellyfishDashboard />

## Setup

### Create React Native / Expo project

Firstly create a brand new project.

<Tabs>
  <TabItem value="react-native" label="React Native">

```bash
npx react-native@latest init JellyfishDashboard
```

  </TabItem>
  <TabItem value="expo-bare" label="Expo Bare workflow">

```bash
npx react-native init jellyfish-dashboard --template react-native-template-typescript
```

  </TabItem>
</Tabs>

### Add dependencies

Please make sure to install or update `expo` to version `^49.0.0` 

<Tabs>
  <TabItem value="react-native" label="React Native">
    In order for this module to work you'll need to also add `expo` package. The
    expo package has a small footprint and it's necessary as Jellyfish Client
    package is built as Expo module.

```bash npm2yarn
npx install-expo-modules@latest
npm install @jellyfish-dev/react-native-client-sdk
```

  </TabItem>
  <TabItem value="expo-bare" label="Expo Bare workflow">

```bash
npx install-expo-modules@latest
npx expo install @jellyfish-dev/react-native-client-sdk
```

  </TabItem>

</Tabs>

:::tip
Run `pod install` in the /ios directory to install the new pods
:::

> ### Native configuration

Reviewer 2: ```suggestion
### Permissions configuration
```

for camera permission.

```xml title=ios/jellyfish-dashboard/Info.plist
<key>NSCameraUsageDescription</key>
<string> 🙏 🎥 </string>
```

Similarly, for audio there is `NSMicrophoneUsageDescription`.

```xml title=ios/jellyfish-dashboard/Info.plist
<key>NSMicrophoneUsageDescription</key>
<string> 🙏 🎤 </string>
```

For screencast there is more configuration needed, it's described [here](https://github.com/jellyfish-dev/react-native-client-sdk#ios).

We also suggest setting background mode to `audio` so that the app doesn't
disconnect when it's in the background:

```xml title=ios/jellyfish-dashboard/Info.plist
<key>UIBackgroundModes</key>
<array>
  <string>audio</string>
</array>
```

  </TabItem>
   <TabItem value="expo-bare" label="Expo Bare workflow"> 
   
   You have two options here. You can follow configuration instructions for
React Native (Expo Bare workflow is a React Native project after all) or if
you're using `expo prebuild` command to set up native code you can add our Expo
plugin. Just add it to `app.json`:

```json title:app.json
{
  "expo": {
    "name": "example",
    //...
    "plugins": ["@jellyfish-dev/react-native-membrane-webrtc"]
  }
}
```

</TabItem>
</Tabs>

> ## Add components library

Reviewer 2: Put this in the `Setup` section

For your convenience, we've prepared a library with nice-looking components
useful for following this tutorial. Feel free to use standard React Native components or your own components though!

```bash
npx expo install @expo/vector-icons expo-barcode-scanner expo-font @expo-google-fonts/noto-sans @jellyfish-dev/react-native-jellyfish-components @react-navigation/native-stack
```

**_You'll also need to install [Reanimated library (3.3.0)](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/getting-started) and [React Navigation (6.1.7)](https://reactnavigation.org/)_**

:::tip
Run `pod install` in the /ios directory to install the new pods
:::

## Connecting to the server and joining the room

> Our app will consist of two screens. The first one allows the user to type, paste or
> scan the peer token and connect to the room. The second screen shows room
> participants with their video tracks.

Reviewer 2: I am not sure about articles placement but I would keep using future simple 

```suggestion
Our app will consist of two screens. 
The first one (ConnectScreen) will allow a user to type, paste or scan a peer token and connect to the room. 
The second one (RoomScreen) will show room participants with their video tracks.
```

> ### Screens

Reviewer 2: I would put this before `Connecting to the server and joining the room`

> For managing screens I've used React Navigation library, you're free though to
> use another one. Our app will contain two screens: Connect screen and Room
> screen:

Reviewer 2: Keep using `we` so

```suggestion
For managing screens we will use React Navigation library, but feel free to pick whatever suits you. 
```

We already mentioned screens in the previous section so I would put their names there. See previous comment

```tsx title="/App.tsx"
import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import ConnectScreen from "./screens/Connect";
import RoomScreen from "./screens/Room";

const Stack = createNativeStackNavigator();

function App(): JSX.Element {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="Connect" component={ConnectScreen} />
        <Stack.Screen name="Room" component={RoomScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}

export default App;
```

### Connect screen

The UI of the Connect screen consists of a simple text input and a few buttons.
The flow for this screen is simple: the user either copies the peer token from the

> dashboard or scans it with a QR code scanner and presses Connect button. The QR

Reviewer 2: Could we start a new sentence in a new line? This make it easier to track changes and add suggestions

just for convenience.

The code for the UI looks like this:

```tsx title="/screens/Connect.tsx"
import React, { useState } from "react";
import { View, StyleSheet } from "react-native";
import {
  Button,
  TextInput,
  QRCodeScanner,
} from "@jellyfish-dev/react-native-jellyfish-components";
import { NavigationProp } from "@react-navigation/native";

interface ConnectScreenProps {
  navigation: NavigationProp<any>;
}

function ConnectScreen({ navigation }: ConnectScreenProps): JSX.Element {
  const [peerToken, setPeerToken] = useState<string>("");

  return (
    <View style={styles.container}>
      <TextInput
        placeholder="Enter peer token"
        value={peerToken}
        onChangeText={setPeerToken}
      />
      <Button
        onPress={() => {
          /* to be filled */
        }}
        title="Connect"
        disabled={!peerToken}
      />
      <QRCodeScanner onCodeScanned={setPeerToken} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    backgroundColor: "#BFE7F8",
    padding: 24,
    rowGap: 24,
  },
});

export default ConnectScreen;
```

### Connecting to the server

> Once the UI is ready, let's implement connecting to the server.

Reviewer 2: ```suggestion
Once the UI is ready, let's implement the logic responsible for connecting to the server.
```

To start the camera we need to ask the user for permission first. We'll use a
standard React Native module for this:

```tsx title="/screens/Connect.tsx"
import { 
  View,
  StyleSheet,
  type Permission,
  PermissionsAndroid,
  Platform, 
 } from "react-native";

// ...

function ConnectScreen({ navigation }: ConnectScreenProps): JSX.Element {
  // ...

  // highlight-start
  const grantedCameraPermissions = async () => {
    if (Platform.OS === 'ios') return true;
    const granted = await PermissionsAndroid.request(
      PermissionsAndroid.PERMISSIONS.CAMERA as Permission
    );
    if (granted !== PermissionsAndroid.RESULTS.GRANTED) {
      console.error("Camera permission denied");
      return false;
    }
    return true;
  };
  // highlight-end

  const connectToRoom = async () => {
    try {
      await connect(JELLYFISH_URL, peerToken.trim());

      // highlight-start
      if (!(await grantedCameraPermissions())) {
        return;
      }
      // highlight-end
    } catch (e) {
      console.log("Error while connecting", e);
    }
  };

  // ...
}

// ...
```

### Starting the camera

Jellyfish Client provides a handy hook for managing the camera: `useCamera`. It

> can not only start a camera but also toggle, switch between multiple cameras,
> manage camera state and camera track simulcast settings and bandwidth. Also when

Reviewer 2: ```suggestion
Not only it can start a camera but also toggle it, manage its state, simulcast and bandwidth settings, and switch between multiple sources.
```

starting the camera you can provide multiple different settings such as
resolution, quality, and metadata. In this example though we'll simply turn it
on to stream the camera to the dashboard with default settings:

```tsx title="/screens/Connect.tsx"
import {
  useJellyfishClient,
  // highlight-next-line
  useCamera,
} from "@jellyfish-dev/react-native-client-sdk";

// ...

function ConnectScreen({ navigation }: ConnectScreenProps): JSX.Element {
  // ...

  // highlight-next-line
  const { startCamera } = useCamera();

  const connectToRoom = async () => {
    try {
      await connect(JELLYFISH_URL, peerToken.trim());

      if (!(await grantedCameraPermissions())) {
        return;
      }

      // highlight-next-line
      await startCamera();
    } catch (e) {
      console.log("Error while connecting", e);
    }
  };

  // ...
}

// ...
```

### Joining the room

> The last step of connecting to the room would be actually joining the room - so
> that your camera track is visible to other users. To do this simply use the `join` function

Reviewer 2: ```suggestion
The last step of connecting to the room would be actually joining it so
that your camera track is visible to the other users. To do this simply use the `join` function
```

from the `useJellyfishClient` hook.

> You can also provide some user metadata when joining. You can put there whatever

Reviewer 2: ```suggestion
You can also provide some user metadata when joining.
Metadata can be anything and is forwarded to the other participants as is.
In our case, we pass a username.
```

you want, it depends on your business logic. In our example, we provide a user name.

After joining the room we navigate to the next screen: Room screen.

```tsx title="/screens/Connect.tsx"
// ...

function ConnectScreen({ navigation }: ConnectScreenProps): JSX.Element {
  // highlight-next-line
  const { connect, join } = useJellyfishClient();

  const connectToRoom = async () => {
    try {
      await connect(JELLYFISH_URL, peerToken.trim());

      if (!(await grantedCameraPermissions())) {
        return;
      }

      await startCamera();

      // highlight-next-line
      await join({ name: "Mobile RN Client" });
      // highlight-next-line
      navigation.navigate("Room");
    } catch (e) {
      console.log("Error while connecting", e);
    }
  };

  // ...
}

// ...
```

Now the app is ready for the first test. If everything went well you should see
a video from your camera in the front-end dashboard. Now onto the second part:
displaying the streams from other participants.

## Displaying streams from other participants

### Displaying video tracks

The Room screen displays the video from the camera on your device and also

> videos from other participants as well. It also allows the user to exit the call.

Reviewer 2: The Room screen has a couple of responsibilities:
* it displays your own video. Note that it's taken directly from your camera i.e. we don't send it to the JF and get it back so other participants might see you a little bit differently 
* it presents current room state so participants list, their video tiles, etc.
* it allows you to leave a meeting

To get information about all participants (also the local one) in the room use
`usePeers()` hook from Jellyfish Client. The hook returns all the participants
with their ids, tracks and metadata. When a new participant joins or any
participant leaves or anything else changes, the hook updates with the new
information.

To display video tracks Jellyfish Client has a dedicated component for

> displaying a video track: `<VideoRenderer>`. It takes a track id as a prop (it

Reviewer 2: Jellyfish Client comes with a dedicated component for displaying a video track: `<VideoRenderer>`. 
It takes a track id as a prop (...) and, as any other `<View>` in react, a style.
`style` property gives a lot of possibilities  ... You can even animate your track!

or sth like this?

### Connect screen

The UI of the Connect screen consists of a simple text input and a few buttons.

> The flow for this screen is simple: the user either copies the peer token from the

Reviewer 2: I think this was okay
